docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,380 @@
+
+## Considered Options
+
+## Decision Outcome
+
+## Solution
+As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
+### Use cases
+After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:
+
+1. Keeping traffic in single zone (disable all cross-zone traffic)
+2. Load balancing based on priorities within a zone. Example: node → AZ → region
+3. Prefer traffic in single zone with fallback to other zones
+4. Load balance equally to list of zones
+5. When leaving zone, target specific zones first. Example zones physically close to each other
+6. Load balance equally to local dpp and other zone
+7. Exclude zone from load balancing
+8. Load balance equally within a group of zones
+
+### API design
+Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
+Because of this we have decided that we will split our API in two sections: `inZone` and `crossZone`.
+
+Here is the full example of policy:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: ["k8s.io/node", "k8s.io/az"] # (1)
+      crossZone: # (2)
+        rules: # (3)
+          - groups: # (4)
+            - name: "group1" # (5)
+              zones: ["zone-1", "zone-2"]
+            - name: "group2"
+              zones: ["zone-3", "zone-4"]
+          - groups:
+            - zones: ["zone-5"]
+        fallback: # (6)
+          type: Any # (7)
+          exclude: # (8)
+            zones: ["zone-6"]
+      overprovisioningFactor: 120% # (9)
+```
+
+
+(1) In `inZone` section, you configure the priorities of tags used for load balancing. If list is empty or inZone is not specified, load balancing works normally, without taking any priorities into account. 
+
+(2) In `crossZone` section, you configure zones priority.
+
+(3) In `rules` section, you configure list of zone groups priorities, first groups will have the highest load balancing priority and last in order will have the lowest priority.
+
+(4) In `groups` section, you configure logical zone groups. Let's assume `zone-1` and `zone-2` are located in the same datacenter, so you want to load balance equally between them. `Zone-5` is located in another datacenter so it should have lower priority, that is why it is in the second groups entry.
+
+(5) In the `name` section, you can assign a name to the group, which is later used for matching when there are multiple groups. If no group name is provided, the values from the group will apply to all entries in previous groups.
```

Comment:
Forgive me if I'm missing something, this MADR is quite complex, but I can't find the example where this group name is used.